### PR TITLE
Migrate MapboxMetricsReporter to coroutines

### PR DESCRIPTION
**Pull request by Lebedinsky**
## Description

Migrate MapboxMetricsReporter to coroutines. Fix tests. Remove ThreadHandler. [issue](https://github.com/mapbox/navigation-sdks/issues/187)

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Align all metrics background work to how it implemented in a whole SDK - with coroutines.

## Testing

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libnavigation-metrics/src/main/java/com/mapbox/navigation/metrics/MapboxMetricsReporter.kt [hunk 1]
```diff
@@ -72,15 +73,15 @@ object MapboxMetricsReporter : MetricsReporter {
...
+        mainJobController.job.cancelChildren()
     }
 
     override fun addEvent(metricEvent: MetricEvent) {
         metricEvent.toTelemetryEvent()?.let {
             mapboxTelemetry.push(it)
         }
 
-        threadWorker.post {
+        mainJobController.scope.launch {
```
Comment (korshaknn) [inline libnavigation-metrics/src/main/java/com/mapbox/navigation/metrics/MapboxMetricsReporter.kt]: why main thread is used? previous  `threadWorker` used `HandlerThread` to process work, so it was in background

File: libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MainCoroutineRule.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineDispatcher
+import kotlinx.coroutines.test.resetMain
+import kotlinx.coroutines.test.runBlockingTest
+import kotlinx.coroutines.test.setMain
+import org.junit.rules.TestWatcher
+import org.junit.runner.Description
+
+@ExperimentalCoroutinesApi
+class MainCoroutineRule(
```
Comment (LukasPaczos) [inline libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MainCoroutineRule.kt]: This should be reused - https://github.com/mapbox/mapbox-navigation-android/issues/2411. Not blocking the PR though.

File: libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MapboxMetricsReporterTest.kt [hunk 3]
```diff
@@ -34,25 +45,27 @@ class MapboxMetricsReporterTest {
...
-    fun telemetryPushCalledWhenAddValidEvent() {
+    fun telemetryPushCalledWhenAddValidEvent() = coroutineRule.runBlockingTest {
         val mapboxTelemetry = initMetricsReporterWithTelemetry()
         val metricEvent = StubNavigationEvent(NavigationMetrics.ARRIVE)
         val event = metricEvent.toTelemetryEvent()
 
         MapboxMetricsReporter.addEvent(metricEvent)
 
         verify { mapboxTelemetry.push(event) }
+        unmockkObject(ThreadController)
```
Comment (LukasPaczos) [inline libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MapboxMetricsReporterTest.kt]: Can this be done in `@After`?

File: libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MapboxMetricsReporterTest.kt [hunk 4]
```diff
@@ -72,9 +85,13 @@ class MapboxMetricsReporterTest {
     }
 
     private fun initMetricsReporterWithTelemetry(): MapboxTelemetry {
+        val parentJob = SupervisorJob()
+        val testScope = CoroutineScope(parentJob + coroutineRule.testDispatcher)
+        mockkObject(ThreadController)
```
Comment (RingerJK) [inline libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MapboxMetricsReporterTest.kt]: better to move `mockkObject(ThreadController)` to `@Before` and  `unmockkObject` to `@After` func

File: libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MapboxMetricsReporterTest.kt (hunk 3 above)
Comment (Guardiola31337) [reply libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MapboxMetricsReporterTest.kt]: Let's be explicit and only add `mockkObject(ThreadController)` from `initMetricsReporterWithTelemetry` https://github.com/mapbox/mapbox-navigation-android/pull/2415/files#diff-04985ebb0b77c4ed5d8ba7df720949c3R90 (and keep `unmockkObject(ThreadController)`) for the tests needed. What do you think?

File: libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MainCoroutineRule.kt (hunk 2 above)
Comment (Lebedinsky) [reply libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MainCoroutineRule.kt]: Yeah I'm agree
But, let's first merge this PR, and in separate PR we will transform `libtesting-ui` to generic module and use `MainCoroutineRule` across all SDK from there. wdyt?

**Comment by Lebedinsky** [comment]
> Don't forget https://github.com/mapbox/mapbox-navigation-android/pull/2415/files#r370172169 @Lebedinsky 😉
> 
> :shipit:

@LukasPaczos  created proper [issue](https://github.com/mapbox/mapbox-navigation-android/issues/2411), so we have no chance to forget it 😄

File: libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MainCoroutineRule.kt (hunk 2 above)
Comment (LukasPaczos) [reply libnavigation-metrics/src/test/java/com/mapbox/navigation/metrics/MainCoroutineRule.kt]: Sounds good :+1:

**Pull request merged**